gobblin-core/src/main/java/gobblin/publisher/HiveRegistrationPublisher.java

```diff
@@ -92,23 +93,38 @@ public void publishData(Collection<? extends WorkUnitState> states) throws IOExc
     CompletionService<Collection<HiveSpec>> completionService =
         new ExecutorCompletionService<>(this.hivePolicyExecutor);
 
-    //addRuntimeHiveRegistrationProperties(super.state);
-    final HiveRegistrationPolicy policy = HiveRegistrationPolicyBase.getPolicy(super.state);
+    State superState = super.state;
+    int numberOfPathsToRegister = getNumberOfPathsToRegister(states);
 
-    Set<String> pathsToRegister = getUniquePathsToRegister(states);
-    log.info("Number of paths to be registered in Hive: " + pathsToRegister.size());
-    for (final String path : pathsToRegister) {
-      completionService.submit(new Callable<Collection<HiveSpec>>() {
+    // Each state in states is task-level State, while superState is the Job-level State.
+    // Using both State objects to distinguish each HiveRegistrationPolicy so that
+    // they can carry task-level information to pass into Hive Partition and its corresponding Hive Table.
 
-        @Override
-        public Collection<HiveSpec> call() throws Exception {
-          return policy.getHiveSpecs(new Path(path));
+    // Here all runtime task-level props are injected into superstate which installed in each Policy Object.
+    // runtime.props are comma-separated props collected in runtime.
+    for (State state:states) {
+      if (state.contains(ConfigurationKeys.PUBLISHER_DIRS)) {
+        if (state.contains(HiveConstants.TOPIC_NAME)) {
```

Comment:
This is still overly specific. Any reference to TOPIC should only be on the Kafka side of things.